db/version_set.cc

```diff
@@ -330,94 +330,69 @@ Status Version::Get(const ReadOptions& options, const LookupKey& k,
-          tmp.push_back(f);
-        }
-      }
-      if (tmp.empty()) continue;
+    VersionSet *vset;
+    Status s;
 
-      std::sort(tmp.begin(), tmp.end(), NewestFirst);
-      files = &tmp[0];
-      num_files = tmp.size();
-    } else {
-      // Binary search to find earliest index whose largest key >= ikey.
-      uint32_t index = FindFile(vset_->icmp_, files_[level], ikey);
-      if (index >= num_files) {
-        files = nullptr;
-        num_files = 0;
-      } else {
-        tmp2 = files[index];
-        if (ucmp->Compare(user_key, tmp2->smallest.user_key()) < 0) {
-          // All of "tmp2" is past any data for user_key
-          files = nullptr;
-          num_files = 0;
-        } else {
-          files = &tmp2;
-          num_files = 1;
-        }
-      }
-    }
+    static bool Match(void* arg, int level, FileMetaData* f) {
+      State* state = reinterpret_cast<State*>(arg);
 
-    for (uint32_t i = 0; i < num_files; ++i) {
-      if (last_file_read != nullptr && stats->seek_file == nullptr) {
+      if (state->stats->seek_file == nullptr) {
         // We have had more than one seek for this read.  Charge the 1st file.
-        stats->seek_file = last_file_read;
-        stats->seek_file_level = last_file_read_level;
+        state->stats->seek_file = f;
+        state->stats->seek_file_level = level;
       }
 
-      FileMetaData* f = files[i];
-      last_file_read = f;
-      last_file_read_level = level;
-
       Saver saver;
       saver.state = kNotFound;
-      saver.ucmp = ucmp;
-      saver.user_key = user_key;
-      saver.value = value;
-      s = vset_->table_cache_->Get(options, f->number, f->file_size, ikey,
-                                   &saver, SaveValue);
+      saver.ucmp = state->ucmp;
+      saver.user_key = state->user_key;
+      saver.value = state->value;
+
+      Status s = state->vset->table_cache_->Get(*state->options, f->number,
+                                                f->file_size, state->ikey,
+                                                &saver, SaveValue);
       if (!s.ok()) {
-        return s;
+        state->s = s;
+        return false;
       }
       switch (saver.state) {
         case kNotFound:
-          break;  // Keep searching in other files
+          return true; // Keep saerching in other files
         case kFound:
-          return s;
+          state->s = s;
+          return false;
         case kDeleted:
-          s = Status::NotFound(Slice());  // Use empty error message for speed
-          return s;
+          return false;
         case kCorrupt:
-          s = Status::Corruption("corrupted key for ", user_key);
-          return s;
+          state->s = Status::Corruption("corrupted key for ", state->user_key);
+          return false;
       }
     }
-  }
+  };
+
+  stats->seek_file = nullptr;
+  stats->seek_file_level = -1;
+
+  State state;
+  state.s = Status::NotFound(Slice());
```

Comment:
Non-OK status values are somewhat expensive.  I suggest adding a "bool found" to State, and if after everything is done without an error and found is false, create the NotFound status at that point.

Reply:
I cache the Saver object in State to avoid creating Saver for each overlapping file(though it may cause extra expense if there is no overlapping files) and the bench result is:
fillseq      :       8.472 micros/op;   13.1 MB/s
fillsync     :    3992.071 micros/op;    0.0 MB/s (1000 ops)
fillrandom   :      11.059 micros/op;   10.0 MB/s
overwrite    :      12.333 micros/op;    9.0 MB/s
readrandom   :       5.943 micros/op; (1000000 of 1000000 found)
readrandom   :       5.396 micros/op; (1000000 of 1000000 found)
readseq      :       0.219 micros/op;  505.2 MB/s
readreverse  :       0.350 micros/op;  315.8 MB/s
compact      :  871343.000 micros/op;
readrandom   :       4.140 micros/op; (1000000 of 1000000 found)
readseq      :       0.177 micros/op;  625.3 MB/s
readreverse  :       0.310 micros/op;  357.2 MB/s
fill100K     :    1347.276 micros/op;   70.8 MB/s (1000 ops)
crc32c       :       1.395 micros/op; 2799.8 MB/s (4K per op)
snappycomp   :       4.086 micros/op;  956.0 MB/s (output: 55.1%)
snappyuncomp :       0.628 micros/op; 6222.3 MB/s